server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -129,11 +134,38 @@ protected void doClose() {
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
         try {
+            final var localNodeId = clusterService.localNode().getId();
+
+            // Track when this node enters and leaves shutdown mode because we pause shard snapshots for shutdown.
+            // The snapshotShutdownProgressTracker will report (via logging) on the progress shard snapshots make
+            // towards either completing (successfully or otherwise) or pausing.
+            NodesShutdownMetadata currentShutdownMetadata = event.state().metadata().custom(NodesShutdownMetadata.TYPE);
+            NodesShutdownMetadata previousShutdownMetadata = event.previousState().metadata().custom(NodesShutdownMetadata.TYPE);
+            SingleNodeShutdownMetadata currentLocalNodeShutdownMetadata = currentShutdownMetadata != null
+                ? currentShutdownMetadata.get(localNodeId)
+                : null;
+            SingleNodeShutdownMetadata previousLocalNodeShutdownMetadata = previousShutdownMetadata != null
+                ? previousShutdownMetadata.get(localNodeId)
+                : null;
+
+            boolean isLocalNodeAddingShutdown = false;
+            if (previousLocalNodeShutdownMetadata == null && currentLocalNodeShutdownMetadata != null) {
+                snapshotShutdownProgressTracker.onClusterStateAddShutdown();
+                isLocalNodeAddingShutdown = true;
+            }
+
+            if (currentLocalNodeShutdownMetadata == null && previousLocalNodeShutdownMetadata != null) {
+                snapshotShutdownProgressTracker.onClusterStateRemoveShutdown();
+            }
```

Comment:
I originally used SnapshotsInProgress#isNodeIdForRemoval, but it turned out to be unpredictable because that field gets cleared without cluster state changes -- IIRC, when the snapshots finish pausing, or something. I pivoted to looking at the shutdown marker as more predictable. But in theory, I think I could use isNodeIdForRemoval, I think 🤔  I shall think more on this.